txnsync/incoming.go

```diff
@@ -0,0 +1,307 @@
+	_, err = incomingMessage.message.UnmarshalMsg(message)
+	if err != nil {
+		// if we received a message that we cannot parse, disconnect.
+		s.log.Infof("received unparsable transaction sync message from peer. disconnecting from peer.")
+		return err
+	}
+
+	if incomingMessage.message.Version != txnBlockMessageVersion {
+		// we receive a message from a version that we don't support, disconnect.
+		s.log.Infof("received unsupported transaction sync message version from peer. disconnecting from peer.")
+		return errUnsupportedTransactionSyncMessageVersion
+	}
+
+	// if the peer sent us a bloom filter, decode it
+	if incomingMessage.message.TxnBloomFilter.BloomFilterType != 0 {
+		bloomFilter, err := decodeBloomFilter(incomingMessage.message.TxnBloomFilter)
+		if err != nil {
+			s.log.Infof("Invalid bloom filter received from peer : %v", err)
+			return errInvalidBloomFilter
+		}
+		incomingMessage.bloomFilter = bloomFilter
+		// increase number of decoded bloom filters.
+		txsyncDecodedBloomFiltersTotal.Inc(nil)
+	}
+
+	// if the peer sent us any transactions, decode these.
+	incomingMessage.transactionGroups, err = decodeTransactionGroups(incomingMessage.message.TransactionGroups, s.genesisID, s.genesisHash)
+	if err != nil {
+		s.log.Infof("failed to decode received transactions groups: %v\n", err)
+		return errDecodingReceivedTransactionGroupsFailed
+	}
+
+	if peer == nil {
+		// if we don't have a peer, then we need to enqueue this task to be handled by the main loop since we want to ensure that
+		// all the peer objects are created synchronously.
+		enqueued := s.incomingMessagesQ.enqueue(incomingMessage)
+		if !enqueued {
+			// if we can't enqueue that, return an error, which would disconnect the peer.
+			// ( we have to disconnect, since otherwise, we would have no way to syncronize the sequence number)
+			s.log.Infof("unable to enqueue incoming message from a peer without txsync allocated data; incoming messages queue is full. disconnecting from peer.")
+			return errTransactionSyncIncomingMessageQueueFull
+		}
+		return nil
+	}
+	// place the incoming message on the *peer* heap, allowing us to dequeue it in the correct sequence order.
+	err = peer.incomingMessages.enqueue(incomingMessage)
+	if err != nil {
+		// if the incoming message queue for this peer is full, disconnect from this peer.
+		s.log.Infof("unable to enqueue incoming message into peer incoming message backlog. disconnecting from peer.")
+		return err
+	}
+
+	// (maybe) place the peer message on the main queue. This would get skipped if the peer is already on the queue.
+	enqueued := s.incomingMessagesQ.enqueue(incomingMessage)
+	if !enqueued {
+		// if we can't enqueue that, return an error, which would disconnect the peer.
+		s.log.Infof("unable to enqueue incoming message from a peer with txsync allocated data; incoming messages queue is full. disconnecting from peer.")
+		return errTransactionSyncIncomingMessageQueueFull
+	}
```

Comment:
This block is repeated once here and once for `peer == nil`
Why not combine them, and run `peer.incomingMessages.enqueue` before this only if peer != nil?

Reply:
these indicates two different issues : 
one is that the peer queue was getting too long.
the other is that we have too many peers connected to the relay.